design-system/pkg/src/split-view/SplitView.tsx

```diff
@@ -0,0 +1,504 @@
+import {
+  classNames,
+  css,
+  tokenSchema,
+  transition,
+  useStyleProps,
+} from '@keystar/ui/style';
+import { useId } from '@keystar/ui/utils';
+import { useLocale } from '@react-aria/i18n';
+import { filterDOMProps, useUpdateEffect } from '@react-aria/utils';
+import { clamp } from '@react-stately/utils';
+import {
+  ForwardedRef,
+  ForwardRefExoticComponent,
+  HTMLAttributes,
+  Ref,
+  forwardRef,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+
+import { SplitViewProvider, useSplitView } from './context';
+import { resetGlobalCursorStyle, setGlobalCursorStyle } from './cursor';
+import { defaultStorage } from './storage';
+import {
+  CursorState,
+  ResizeEvent,
+  SplitPanePrimaryProps,
+  SplitPaneSecondaryProps,
+  SplitViewProps,
+} from './types';
+import {
+  getPercentage,
+  getPosition,
+  getPrimaryPane,
+  getPrimaryPaneId,
+  getResizeHandle,
+  getResizeHandleId,
+  getSecondaryPaneId,
+  px,
+} from './utils';
+
+const MAX_WIDTH_PROP = '--primary-pane-max-width';
+const MAX_WIDTH_VAR = `var(${MAX_WIDTH_PROP})`;
+const MIN_WIDTH_PROP = '--primary-pane-min-width';
+const MIN_WIDTH_VAR = `var(${MIN_WIDTH_PROP})`;
+const WIDTH_PROP = '--primary-pane-width';
+const WIDTH_VAR = `var(${WIDTH_PROP})`;
+const SNAP_REGION_PX = 32;
+const KEYBOARD_ARROW_STEPS = 10;
+
+export function SplitView(props: SplitViewProps) {
+  let {
+    autoSaveId,
+    children,
+    defaultSize,
+    isCollapsed,
+    minSize,
+    maxSize,
+    onCollapseChange,
+    onResize,
+    storage = defaultStorage,
+  } = props;
+  const [startPane, endPane] = children;
+
+  const id = useId(props.id);
+  const { direction } = useLocale();
+  const styleProps = useStyleProps(props);
+
+  const [isDragging, setDragging] = useState(false);
+  // const [collapseRequested, setCollapseRequested] = useState(false);
+  const [handleIsFocused, setHandleFocus] = useState(false);
+  const [size, setSize] = useState(() => {
+    let size = defaultSize;
+    if (autoSaveId) {
+      let savedSize = storage.getItem(autoSaveId);
+      if (savedSize) {
+        size = Number.parseInt(savedSize);
+      }
+    }
+    return size;
+  });
+
+  const wrapperRef = useRef<HTMLDivElement>(null);
+  const offsetRef = useRef(0);
+  const moveRef = useRef(0);
+
+  // reverse drag logic when the primary pane is on the right or if the locale
+  // direction is right-to-left
+  const isReversed = useMemo(() => {
+    // @ts-expect-error FIXME: this feels super dodgy
```

Comment:
tbh, personally I would forgo trying to do the compositional api thing here of seperate components and just make it one thing (not a blocking thing, just a thought if it's helpful)

Reply:
resolved by ec140e08a0b67b9dbad8eaceaf5ba97f6ac8f3f0